File: cpp/src/11.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cassert>

using namespace std;


class Solution {
public:
    int maxArea(vector<int>& height) {
        auto left = height.begin(), right = height.end()-1;
        int max_area = 0;
        
        while (left < right) {
            if (min(*left, *right) * (right-left) > max_area) {
                max_area = min(*left, *right) * (right-left);
            }
            
            if (*left < *right) left++;
            else right--;
        }
        
        return max_area;
    }
};


int main () {
    vector<int> in = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    Solution solution = Solution();
    int max_area = solution.maxArea(in);

    assert(max_area == 49);
    cout << "Success" << endl;

    return 0;
}

File: go/src/11.go
func maxArea(height []int) int {
	i := 0
	j := len(height) - 1

	max := 0

	for i < j {
		curVol := min(height[i], height[j]) * (j - i)
		if max < curVol {
			max = curVol
		}
		if height[i] <= height[j] {
			i++
		} else {
			j--
		}
	}

	return max
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}